### PR TITLE
Thread id bug

### PR DESCRIPTION
**Pull request by Leo10Gama**
#### Which issue(s) does this change fix?
N/A

#### Why is this change necessary?
Previously, Python's `threading.thread_id()` method did not provide an accurate representation of distinct Events running on the same thread. This change allows for a more accurate look at which methods are running on the same thread, and helps group long Events more accurately.

#### How does it address the issue?
The previous `Event.thread_id` was changed from an `int` to a `UUID` to act as a better identifier for Events run on the same thread.

#### What side effects does this change have?
The `thread_id` will now be a UUID, and its representation in JSON and `__repr__` formats will be its hex value, as a string.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [x] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [x] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/telemetry/event.py [hunk 1]
```diff
@@ -144,7 +150,11 @@ class EventTracker:
 
     @staticmethod
     def track_event(
-        event_name: str, event_value: str, session_id: Optional[str] = None, exception_name: Optional[str] = None
+        event_name: str,
+        event_value: str,
+        session_id: Optional[str] = None,
+        thread_id: Optional[UUID] = None,
```
Comment (hawflau) [inline samcli/lib/telemetry/event.py:156]: does putting `thread_id` before `exception_name` as positional argument affect any existing `track_event` call?

**Review by hawflau — COMMENTED** [review]
overall LGTM. Just one question - can you also check if some change is needed for the `track_event` calls in samcli/lib/sync/infra_sync_executor.py?

File: samcli/lib/telemetry/event.py (hunk 1 above)
Comment (Leo10Gama) [reply samcli/lib/telemetry/event.py:156]: Nope it does not, just verified in the code base. The only call right now that makes use of the `session_id` property is the call in `local_apigw_service.py` on line 701, which calls it by keyword.

File: samcli/lib/sync/infra_sync_executor.py [hunk 2]
```diff
@@ -174,8 +175,9 @@ def execute_infra_sync(self, first_sync: bool = False) -> InfraSyncResult:
             days_since_last_infra_sync = (current_time - last_infra_sync_time).days
 
         # Will not combine the comparisons in order to save operation cost
+        thread_id = uuid4()
         if self._sync_context.skip_deploy_sync and first_sync and (days_since_last_infra_sync <= AUTO_INFRA_SYNC_DAYS):
-            EventTracker.track_event("SyncFlowStart", "SkipInfraSyncExecute")
+            EventTracker.track_event("SyncFlowStart", "SkipInfraSyncExecute", thread_id=thread_id)
```
Comment (mndeveci) [inline samcli/lib/sync/infra_sync_executor.py:180]: I think you missed to add `thread_id` to line 191

**Review by mndeveci — CHANGES_REQUESTED** [review]
LGTM in general, left some comments there!

File: samcli/lib/telemetry/event.py [hunk 3]
```diff
@@ -83,14 +84,19 @@ class Event:
...
 
-    def __init__(self, event_name: str, event_value: str, exception_name: Optional[str] = None):
+    def __init__(
+        self, event_name: str, event_value: str, thread_id: Optional[UUID] = None, exception_name: Optional[str] = None
+    ):
         Event._verify_event(event_name, event_value)
         self.event_name = EventName(event_name)
         self.event_value = event_value
+        if not thread_id:
+            thread_id = uuid4()
```
Comment (mndeveci) [inline samcli/lib/telemetry/event.py:98]: Should we still use `threading.get_ident()` for default value or other fields will same for that purpose?
@hawflau what do you think?

File: samcli/lib/sync/infra_sync_executor.py (hunk 2 above)
Comment (Leo10Gama) [reply samcli/lib/sync/infra_sync_executor.py:180]: Yup I missed that, good catch

File: samcli/lib/telemetry/event.py (hunk 3 above)
Comment (hawflau) [reply samcli/lib/telemetry/event.py:98]: I think `uuid4()` is good. From what I've seen empirically, `session_id` and `threading.get_ident()` are producing the same grouping effects.

**Review by mndeveci — APPROVED** [review]

**Review by hawflau — APPROVED** [review]

**Pull request merged**